File: hphp/hack/src/naming/naming_types.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

type kind_of_type =
  | TClass
  | TTypedef
  | TRecordDef
[@@deriving show, eq]

type name_kind =
  | Type_kind of kind_of_type
  | Fun_kind
  | Const_kind
[@@deriving show, eq]

let type_kind_to_name_type (kind_of_type : kind_of_type) : FileInfo.name_type =
  match kind_of_type with
  | TClass -> FileInfo.Class
  | TRecordDef -> FileInfo.RecordDef
  | TTypedef -> FileInfo.Typedef

let name_kind_to_name_type (name_kind : name_kind) : FileInfo.name_type =
  match name_kind with
  | Type_kind type_kind -> type_kind_to_name_type type_kind
  | Fun_kind -> FileInfo.Fun
  | Const_kind -> FileInfo.Const

let name_kind_of_name_type (name_type : FileInfo.name_type) : name_kind =
  match name_type with
  | FileInfo.Class -> Type_kind TClass
  | FileInfo.Typedef -> Type_kind TTypedef
  | FileInfo.RecordDef -> Type_kind TRecordDef
  | FileInfo.Fun -> Fun_kind
  | FileInfo.Const -> Const_kind

let type_kind_of_name_type (name_type : FileInfo.name_type) :
    kind_of_type option =
  match name_type with
  | FileInfo.Class -> Some TClass
  | FileInfo.Typedef -> Some TTypedef
  | FileInfo.RecordDef -> Some TRecordDef
  | FileInfo.Fun
  | FileInfo.Const ->
    None

let name_kind_to_enum (name_kind : name_kind) : int =
  (* These values are stored in the naming table database, and must match [flag_of_enum] and naming_types_impl.rs
     We happen for convenience to use the same numerical values as FileInfo.name_type. *)
  name_kind |> name_kind_to_name_type |> FileInfo.name_type_to_enum

let name_kind_of_enum (i : int) : name_kind option =
  (* These values are stored in the naming table database, and must match [flag_to_enum] and naming_types_impl.rs *)
  i |> FileInfo.name_type_of_enum |> Option.map name_kind_of_name_type

File: hphp/hack/src/naming/rust/naming_types.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
//
// @generated <<SignedSource::*O*zOeWoEQle#+L!plEphiEmie@IsG>>
//
// To regenerate this file, run:
//   buck run @fbcode//mode/dev-nosan-lg fbcode//hphp/hack/src:oxidized_regen
use arena_trait::TrivialDrop;
use eq_modulo_pos::EqModuloPos;
use no_pos_hash::NoPosHash;
use ocamlrep::FromOcamlRep;
use ocamlrep::FromOcamlRepIn;
use ocamlrep::ToOcamlRep;
use serde::Deserialize;
use serde::Serialize;

#[allow(unused_imports)]
use crate::*;

/// In the naming table, global constants and global functions can be
/// syntactically disambiguated at their use-site, and therefore can live in
/// separate namespaces. However, types (such as classes) cannot be syntactically
/// disambiguated, and they live in the same namespace. So in the naming table,
/// we also have to store what kind of type that symbol was.
#[derive(
    Clone,
    Copy,
    Debug,
    Deserialize,
    Eq,
    EqModuloPos,
    FromOcamlRep,
    FromOcamlRepIn,
    Hash,
    NoPosHash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
    ToOcamlRep,
)]
#[rust_to_ocaml(attr = "deriving (show, eq)")]
#[repr(u8)]
pub enum KindOfType {
    TClass,
    TTypedef,
}
impl TrivialDrop for KindOfType {}
arena_deserializer::impl_deserialize_in_arena!(KindOfType);

#[derive(
    Clone,
    Debug,
    Deserialize,
    Eq,
    EqModuloPos,
    FromOcamlRep,
    Hash,
    NoPosHash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
    ToOcamlRep,
)]
#[rust_to_ocaml(attr = "deriving (show, eq)")]
#[repr(C, u8)]
pub enum NameKind {
    #[rust_to_ocaml(name = "Type_kind")]
    TypeKind(KindOfType),
    #[rust_to_ocaml(name = "Fun_kind")]
    FunKind,
    #[rust_to_ocaml(name = "Const_kind")]
    ConstKind,
    #[rust_to_ocaml(name = "Module_kind")]
    ModuleKind,
}
